test/conftest.py

```diff
@@ -376,8 +377,17 @@ def requires_tlsv1_3(supported_tls_versions: typing.AbstractSet[str]) -> None:
 def http_version(request: pytest.FixtureRequest) -> typing.Generator[str, None, None]:
     if request.param == "h2":
         urllib3.http2.inject_into_urllib3()
+    try:
+        yield request.param
+    finally:
+        if request.param == "h2":
+            urllib3.http2.extract_from_urllib3()
 
-    yield request.param
 
-    if request.param == "h2":
-        urllib3.http2.extract_from_urllib3()
+@pytest.fixture(autouse=True)
```

Comment:
Would it make sense to be more explicit than `autouse`, and make `reset_http2_probe_cache` a dependency of the `http_version` fixture?

This means you'll need to add it to `test_http2_probe_no_result_in_connect_error` and `test_http2_probe_no_result_in_ssl_error`, which makes sense anyway.

Reply:
I'm going to keep this as-is, since even in this current PR the HTTP/2 probe cache is used in tests that don't use the `http_version` fixture.

Reply:
OK, then I have a final nit:

```suggestion
@pytest.fixture(scope="function", autouse=True)
```